### PR TITLE
Fix regression introduced by fix for Symfony 3, check for undefined schemes key

### PR DESCRIPTION
**Pull request by tobias-93**
Fix for #287, problem introduced by #254

### CODE REVIEW

**Comment by GuilhemN** [comment]
Any Idea why the tests are failing?

**Comment by tobias-93** [comment]
@GuilhemN I think the tests are a little more strict than browsers are. I just added another check which should not trigger the test to fail.

**Comment by tobias-93** [comment]
Ok, I'll have to look into them, as they still fail

**Comment by willdurand** [comment]
I see a lot of:

```
TypeError: 'undefined' is not an object (evaluating 'route.schemes.length') 'undefined' is not an object (evaluating 'route.schemes.length') TypeError: 'undefined' is not an object (evaluating 'route.schemes.length')
```

**Comment by maguedon** [comment]
I think it is because the property "schemes" may not exist in some cases. You should check if it exists first. Why did you remove the condition `goog.object.containsKey(route, "schemes")` ? I suppose it checks if "schemes" exists. Does it work if you add it at the beginning of your condition like this :

``` js
else if (goog.object.containsKey(route, "schemes") && route.schemes.length > 0 && typeof route.schemes[0] !== "undefined" && this.getScheme() != route.schemes[0]) {
```

**Comment by tobias-93** [comment]
I forgot to add the new object keys to the JS tests, so it should be fixed now

**Comment by tobias-93** [comment]
The tests now work again. @maguedon that would also fix it, but we can assume this key will always exist since the Symfony command always adds it to the object.

**Comment by willdurand** [comment]
> The tests now work again. @maguedon that would also fix it, but we can assume this key will always exist since the Symfony command always adds it to the object.

Is that 100% sure? For BC purpose I would rather go for a test and not assume it exists..

**Comment by tobias-93** [comment]
@willdurand It should be. I added the check anyway, just to be sure.

File: Resources/js/router.js [hunk 1]
```diff
@@ -241,7 +241,7 @@ fos.Router.prototype.generate = function(name, opt_params, absolute) {
     url = this.context_.base_url + url;
     if (goog.object.containsKey(route.requirements, "_scheme") && this.getScheme() != route.requirements["_scheme"]) {
         url = route.requirements["_scheme"] + "://" + (host || this.getHost()) + url;
-    } else if (goog.object.containsKey(route, "schemes") && this.getScheme() != route.schemes[0]) {
+    } else if (goog.object.containsKey(route, "schemes") && route.schemes.length > 0 && typeof route.schemes[0] !== "undefined" && this.getScheme() != route.schemes[0]) {
```
Comment (GuilhemN) [inline Resources/js/router.js]: I don't think we need both checks :) (`.length` and `!== undefined`)

Comment (tobias-93) [inline Resources/js/router.js]: I agree, doesn't do any harm though

**Comment by GuilhemN** [comment]
Is this ok to merge @willdurand ?

**Comment by willdurand** [comment]
It would be better not to update the whole test suite to ensure BC but otherwise sure :+1:

**Comment by tobias-93** [comment]
@willdurand I have removed the updates to the test suite, so just the new tests remain.
@GuilhemN I have removed the duplicate check.
Is this ready to be merged now?

**Comment by willdurand** [comment]
Looks good to me!

**Comment by willdurand** [comment]
Looks good to me!

**Pull request merged**